policy: Reduce allocs when keeping track of owners

jrajahalme opened the pull request:
Store single owners directly in the `MapStateEntry` and switch to a map only when more than one owner is needed. This is done with the new `ifset.Set`.

Instead of a map of MapStateOwner interfaces, the Owners member of MapStateEntry is now a `ifset.Set`, which can hold a single interface value directly, or a more in a map. The new `ifset.Set` type uses an unexported `memberMap` type for its internal map, so the Set can hold any interface values (including empty interface value `nil`).

With this change the `BenchmarkRegenerateCIDRDenyPolicyRules` benchmark runs ~11% faster and uses ~15% less memory with ~9% less allocs than with the prior map only version.

jrajahalme committed:
policy: Unexport MapStateEntry.Merge

MapStateEntry.Merge need not be exported.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Only delete mapstate entry from the map it exists in

MapState keeps separate maps for allows and denies. There is no need to
delete a key with a known entry value from both maps, as we can observe
if the entry is a delete or an allow entry.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Reduce allocs when keeping track of owners

Use the new 'set.Set' to store owners in MapStateEntry.

Extend MapStateEntry.String() to output the owners, dependents, and
listener priority. This eases unit test debugging going forward.

With this change the BenchmarkRegenerateCIDRDenyPolicyRules benchmark
runs ~10% faster and uses ~15% less memory with ~9% less allocs than with
the prior map only version.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

nebril commented on pkg/container/ifset/set.go:
```diff
@@ -0,0 +1,378 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
```
nitpick: Why Hubble?

jrajahalme replied:
Must have been copy&paste from a file that said Hubble :-) Fixed!

nebril commented on pkg/policy/mapstate_test.go:
```diff
@@ -2147,6 +2226,7 @@ func TestMapState_AccumulateMapChanges(t *testing.T) {
 	policyMapState := newMapState()
 
 	for _, tt := range tests {
+		fmt.Println(tt.name)
```
Is this a leftover?

jrajahalme replied:
Yes! Removed!

nebril reviewed (approved):
Nice! Two nitpicks left inline.

pippolo84 commented on pkg/container/ifset/set.go:
```diff
@@ -0,0 +1,378 @@
+// Member is an empty interface to hold values in Set
+type Member interface {
+	String() string
+}
```
This is just a duplicate of the [fmt.Stringer](https://pkg.go.dev/fmt#Stringer) interface.

Anyway, I think we should avoid constraining the types to be used for a Set with this interface. Instead, I'd rather use `any` for broader reusability of the data structure and leave the responsibility of stringify the elements to the user, whenever possible (e.g: for the `mapstate` package we can write a function that first assert the type and then produce a string accordingly).

jrajahalme replied:
Right, added this for debugging and did not think at that point this limits the member types we can use in a set! Will change back to `any`!

jrajahalme replied:
Now with generics the type constraint is `comparable`.

pippolo84 commented on pkg/container/ifset/set.go:
```diff
@@ -0,0 +1,378 @@
+}
+
+// ForEachOfType return an iterator for each member of type T in the set.
+func ForEachOfType[T any](s Set) iter.Seq[T] {
```
AFAIU we need this data structure to hold heterogeneous types becase a `MapStateOwner` can be a`policyTypes.Key` or a `policy.CachedSelector`. Moreover we need it to be nullable.

Is that correct or am I missing any other types?

My hunch is that we are trying to deal with some tech debt on the policy side (namely in the way we are representing the owner of a map state entry) with a very specialized data structure for which I see very little opportunity to be reused.

Would it be possible to declare a single interface that describes all the possible types of owners of a map state entry? If so, we could use it to instantiate a `Set` like this one for the map state `owners`:

```go
type Set[T comparable] struct {
   ...
}
```

What the cost of that refactoring would be on the policy code?

Compared to the current implementation,I think  there will be multiple benefits with a Set like this one (broader reusability of the Set, easier implementation, better help from the compiler due to the type parameter, and so on).

jrajahalme replied:
I had this first in `pkg/policy/types`, but thought maybe there is a chance that we need sets of heterogeneous interfaces elsewhere as well, so moved this to `pkg/containers/ifset` for this PR.

I'll see how this would work out using generics instead like you propose here.

jrajahalme replied:
@pippolo84 Done. Using generics it is not possible to use the same struct member for the single element of the type `T` and for the `map[T]`. But this also simplifies the implementation quite a bit!

Also ended by by referring to the singular element via a pointer, as then we do not need to add a separate (boolean) member to tell if the singular member is occupied or not.

In the `BenchmarkRegenerateCIDRDenyPolicyRules` this performs more or less the same as the previous version.

Since the member can now be any `comparable` type, I moved the implementation to `pkg/container/set`.

pippolo84 reviewed (commented):
Did a first pass and currently trying to better understand the scope of the new Set data structure. Left some questions inline.

pippolo84 commented on pkg/container/set/set.go:
```diff
@@ -0,0 +1,268 @@
+	"maps"
+)
+
+type Empty struct{}
```
No need to export `Empty` outside of package set.

jrajahalme replied:
unexported :-)

pippolo84 commented on pkg/container/set/set.go:
```diff
@@ -0,0 +1,268 @@
+// so that if it is stored by value the caller must knows to update the stored value.
+// If 's' holds a map and new members are added, 'false' will be returned,
+// as the value 's' is not changed in this case.
+func (s *Set[T]) Insert(member T) (changed bool) {
```
I'd rather let the `Insert` method add a single value only.

Instead we can add a `Merge` method to handle the insertion of one or more sets:

```go
func Merge[T comparable](sets ...Set[T]) Set[T] {
	set := Set[T]{}

	for _, other := range sets {
		for value := range other.Members() {
			set.Insert(value)
		}
	}

	return set
}
```

This way we can avoid the type assertions.

jrajahalme replied:
I kept it a bit simpler with a `*Set` being a receiver.

pippolo84 commented on pkg/container/set/set.go:
```diff
@@ -0,0 +1,268 @@
+// so that if it is stored by value the caller knows to update the stored value.
+// If 's.members' remains a map, 'false' will be returned,
+// as the value 's' is not changed in this case.
+func (s *Set[T]) Remove(member T) (found, changed bool) {
```
Same for `Remove`, I'd rather keep it simple and maybe add something like:

```go
func Diff[T comparable](first Set[T], others ...Set[T]) Set[T] {
	res := first.Clone()

	for _, other := range others {
		for value := range other.Members() {
			res.Remove(value)
		}
	}

	return res
}
```

to handle the removal of one or more sets from the first one.

jrajahalme replied:
I kept it a bit simpler with a *Set being a receiver, and naming this `RemoveSets()`. Caller can do the cloning if needed.

pippolo84 commented on pkg/container/set/set.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+// Get any item of type M from the set
+func Get[M any, T comparable](s Set[T]) (m M, found bool) {
```
This is actually unused.
I'd rather remove it since the case where we have multiple heterogeneous types in the same set is something that we should avoid as much as possible (it basically defeats the point of having the type parameter).

jrajahalme replied:
Thanks, I'm using `MembersOfType[Key]` iterator where I was using this originally, removed!

pippolo84 commented on pkg/container/set/set.go:
```diff
@@ -0,0 +1,268 @@
+		m = *s.single
+	default:
+		for m = range s.members {
+		}
```
Nit:

```go
		for m = range maps.Keys(s.members) {
			break
		}
```

Since we need the first random key in the map, we might as well break after the first iteration to improve performance.

jrajahalme replied:
done!

pippolo84 commented on pkg/policy/mapstate.go:
```diff
@@ -31,6 +32,7 @@ import (
 // lest ye find yourself with hundreds of unnecessary imports.
 type Key = policyTypes.Key
 type Keys = policyTypes.Keys
+type MapStateOwner = fmt.Stringer
```
I'm still concerned about using `fmt.Stringer` here. If we can't define a new interface that better describes the common traits between all possible map state owners (like `policyTypes.Key` and `policy.CachedSelector`), why not just use `any`?

jrajahalme replied:
The only common trait (in addition to comparing `MapStateOwner`s is the `String()` method used for string output in tests.

jrajahalme replied:
Turns out `any` works, thanks!

pippolo84 commented on pkg/container/set/set.go:
```diff
@@ -0,0 +1,268 @@
+// so that if it is stored by value the caller knows to update the stored value.
+// If 's.members' remains a map, 'false' will be returned,
+// as the value 's' is not changed in this case.
+func (s *Set[T]) Remove(member T) (found, changed bool) {
```
Also, we can avoid the `found` return value as the caller can simply use `Has` before `Remove`.

jrajahalme replied:
Slightly less efficient, but this simplifies the implementation, so changed as suggested :-)

pippolo84 commented on pkg/container/set/set.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+// Get returns any one member from the set.
+func (s Set[T]) Get() (m T, found bool) {
```
Getting the first random key from a set seems a bit unusual.
What about moving this to the sole place we are actually using it (`consumeMapChanges`)? We could also add a check for the actual set size and react somehow if not equal to 1 (log, panic, health report, etc.)

jrajahalme replied:
Caller would need to use an iterator to get access to the "first" element. Changed this by adding this comment on the impelementation:
```
// Useful when it is known that the set has only one element.
```

pippolo84 commented on pkg/policy/mapstate.go:
```diff
@@ -1633,9 +1645,8 @@ func (mc *MapChanges) consumeMapChanges(p *EndpointPolicy, features policyFeatur
-			for cs := range mc.synced[i].Value.owners { // get the sole selector
-				p.policyMapState.deleteKeyWithChanges(mc.synced[i].Key, cs, changes)
-			}
+			cs, _ := mc.synced[i].Value.owners.Get() // get the sole selector
```
[Question] Just out of curiosity, why here we'll always have at most one selector?

jrajahalme replied:
It's due to how `mc` is populated. Multiple owners becomes possible only after entries are added to the MapState.

pippolo84 reviewed (changes requested):
Thanks, this looks better. :pray: 
I've take a deeper look at the last two commits and left some additional feedback.

jrajahalme committed:
containers: Add new set.Set type

Add new generic container type set.Set optimized for cases where there
typically is only one value in the set. A singular member is stored
separately, while sets with more members use a map to store them.

Set is not a refecence value, like a map is, so whenever a Set is stored
by value (e.g., in a map as a value rather than via a pointer), the value
must be updated when the Set changes from holding n/one to multiple and
from multiple to n/one values.

Due to this added complexity the Set should only be used in cases where
there typically is at most one value, and the associated reduction in
allocations is critical for performance or for memory use.

The singular value is stored by reference in order to avoid the need for
an additional boolean field to tell if the singlular value member is
occupied or not. This saves about 10% of memory due to the 'set.Set'
itself being smaller and has roughly the same performance as an inline
value, when tested for interface member types.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

pippolo84 reviewed (approved):
Thanks! :100:
